source/common/stats/thread_local_store.cc

```diff
@@ -61,6 +61,29 @@ std::list<GaugeSharedPtr> ThreadLocalStoreImpl::gauges() const {
   return ret;
 }
 
+std::list<ParentHistogramSharedPtr> ThreadLocalStoreImpl::histograms() const {
+  // Handle de-dup due to overlapping scopes.
+  std::list<ParentHistogramSharedPtr> ret;
+  std::unordered_set<std::string> names;
+  std::unique_lock<std::mutex> lock(lock_);
+  // TODO(ramaraochavali): As histograms don't share storage, there is a chance of duplicate names
+  // here. We need to process global storage for histograms similar to how we have a central storage
+  // in shared memory for counters/gauges.
+  for (ScopeImpl* scope : scopes_) {
+    for (auto name_histogram_pair : scope->central_cache_.histograms_) {
+      const std::string& hist_name = name_histogram_pair.first;
+      const ParentHistogramSharedPtr& parent_hist = name_histogram_pair.second;
+      if (names.insert(hist_name).second) {
+        ret.push_back(parent_hist);
+      } else {
+        ENVOY_LOG(debug, "duplicate histogram {}.{}", scope->prefix_, hist_name);
```

Comment:
Given that this will potentially have an effect on output, should this one actually be a warn? I could go either way on this. It's conceivable this would happen in the data path, though really it should only happen at config load time, so I'm not that worried about log spam.

You might also make the log message something like "duplicate histogram {}.{}: data loss will occur on output".

Can you also open a tracking issue for this? I have ideas on how to fix, but given that this won't effect too many people I don't think it's super urgent (ironically it will for sure effect us at Lyft since we share stats across listeners in certain cases).

Reply:
Changed to warn and added this issue https://github.com/envoyproxy/envoy/issues/3098